### PR TITLE
[TableGen] Add const variants of accessors for backend

### PR DESCRIPTION
**Pull request by jurahul**
Split RecordKeeper `getAllDerivedDefinitions` family of functions into two variants: 
  (a) non-const ones that return vectors of `Record *` and 
  (b) const ones, that return vector/ArrayRef of `const Record *`.

This will help gradual migration of TableGen backends to use 
`const RecordKeeper` and by implication change code to work 
with const pointers and better const correctness.

Existing backends are not yet compatible with the const family of
functions, so change them to use a non-constant `RecordKeeper` 
reference, till they are migrated.

### CODE REVIEW

**Comment by jurahul** [comment]
Note, this is related to the following discourse thread that I had started: https://discourse.llvm.org/t/changing-tablegen-getallderiveddefinitions-to-return-arrayref-const-record/80586/5

File: clang/utils/TableGen/ClangAttrEmitter.cpp [hunk 1]
```diff
@@ -189,7 +189,7 @@ static StringRef NormalizeGNUAttrSpelling(StringRef AttrSpelling) {
 
 typedef std::vector<std::pair<std::string, const Record *>> ParsedAttrMap;
 
-static ParsedAttrMap getParsedAttrList(const RecordKeeper &Records,
```
Comment (matthias-springer) [inline clang/utils/TableGen/ClangAttrEmitter.cpp:192]: I expected to see more `const` instead of less. Why is this here (and below) passes as non-const?

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: Please see the discourse thread: https://discourse.llvm.org/t/changing-tablegen-getallderiveddefinitions-to-return-arrayref-const-record/80586/5

Changing all backends to use const in a single go will be too large a change, so this is a way to stage it. Over time, all backends should move to const variant, and then the non-const ones can be deleted.

Comment (arsenm) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: Would be good to have those conversions queued up to see how this progresses towards that goal

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: I will work on some of them, but not sure if I have the bandwidth to address all of them (Maybe the first few might be more involved and then we will have a better idea of the scope). These will be in similar vein to what I have already done for the IntrinsicEmitter and JSON/Detailed record emitters.

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: I propose the following next steps (once this is committed):

1. We need to send a PSA as current downstream backends may break due to this (fix is easy)
2. I'll work on migrating on of each MLIR. Clang, and may be another LLVM backends to use const.
3. For the rest, we need to chip away. May be we try to recruit volunteers?
4. Before we can deprecate the non-const functions completely, give enough time for downstream backends to change. Maybe this should be a part of the PSA in (1).
5. Delete the non-const functions here, and any other const vs non-const difference we may have introduced during the transition.

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: Here's an example of migrating one MLIR backend: https://github.com/llvm/llvm-project/commit/8ffa5e92662db6aa66a7520da80a360856ac1440

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: And here's for the entire MLIR tablegen code: https://github.com/llvm/llvm-project/commit/061076db7f668f67183e813db6d339a840a05414

So may be its not that involved. Note that this does not address all the const correctness issues, as const Record* function may themselves non-const pointers. So that needs to be fixed gradually as well. So I propose we get this commit in, and then the MLIR one as a follow on. I will see if I can do the clang one as well.

Comment (jurahul) [reply clang/utils/TableGen/ClangAttrEmitter.cpp:192]: And for all clang-tablegen its here: https://github.com/llvm/llvm-project/commit/abf19e5d26ccbf0714fbb93dfc8c783da9e407a1

So looks like we may be able to migrate quickly, may be wait for a few days (weeks?) for downstream adoption and delete the non-const overloads.

**Review by matthias-springer — APPROVED** [review]

**Review by mshockwave — APPROVED** [review]
> I propose the following next steps (once this is committed):

this timeline looks fine, do you also want to add `[[deprecated]]` once the PSA is out?
Also, I can help on other LLVM TableGen backends if they haven't been updated.

**Comment by jurahul** [comment]
> > I propose the following next steps (once this is committed):
> 
> this timeline looks fine, do you also want to add `[[deprecated]]` once the PSA is out? Also, I can help on other LLVM TableGen backends if they haven't been updated.

Yeah, once all the upstream backends are migrated, we should mark the non-const ones deprecated. As for other backends, thanks for the offer! I have diffs for clang and mlir, so I am hoping I can do any other remaining ones as well (lldb and libc seem to have one). But if its start to get delayed, I can ping you for help.

**Comment by jurahul** [comment]
I also did a minor update to the diff: I changed the const versions that used to return references to vectors to instead return `ArrayRef<const Record *>`. IMO this will prevent accidental vector copies in code like `auto X = RC.getAllDerivedDefinitions("yyy")`.

**Comment by jurahul** [comment]
I posted the PSA at: https://discourse.llvm.org/t/psa-planned-changes-to-tablegen-getallderiveddefinitions-api-potential-downstream-breakages/81089
